### PR TITLE
Fix for SYN-266

### PR DESCRIPTION
**Pull request by dbkr**
Does this look sensible?

### CODE REVIEW

File: synapse/storage/registration.py [hunk 1]
```diff
@@ -144,3 +144,21 @@ def _query_for_auth(self, txn, token):
             return rows[0]
 
         raise StoreError(404, "Token not found.")
+
+    @defer.inlineCallbacks
+    def all_users_on_domain(self, domain):
```
Comment (NegativeMjark) [inline synapse/storage/registration.py]: The name made me think it was a way to list all the users on a domain. I guess I'd prefer it to be "are_all_users_on_domain" and to have some docs to say what the method does.

**Comment by NegativeMjark** [comment]
Other than what I've commented on inline it looks good to me.

File: synapse/storage/registration.py (hunk 1 above)
Comment (dbkr) [inline synapse/storage/registration.py]: That would make a lot more sense.

File: synapse/app/homeserver.py [hunk 2]
```diff
@@ -241,6 +242,22 @@ def start_listening(self):
...
+            self.hostname
+        )
+        if not all_users_native:
+            sys.stderr.write(
+                "\n"
+                "******************************************************\n"
+                "Found users in database not native to %s!\n"
+                "You cannot changed a synapse server_name after it's been configured\n"
+                "******************************************************\n"
+                "\n"
```
Comment (NegativeMjark) [inline synapse/app/homeserver.py]: Should there be a % (server_name) here?

Comment (erikjohnston) [inline synapse/app/homeserver.py]: Should this not also log to the log? I don't think we want to assume that stderr goes anywhere sensible once we've daemonized.

File: synapse/app/homeserver.py [hunk 3]
```diff
@@ -399,6 +416,8 @@ def setup(config_options):
     hs.get_datastore().start_profiling()
     hs.get_replication_layer().start_get_pdu_cache()
 
+    reactor.callWhenRunning(hs.post_startup_check)
+
```
Comment (erikjohnston) [inline synapse/app/homeserver.py]: Would it make more sense to check this before daemonizing? This fits in more with the correct situation where bad config causes synapse to die early, allowing people to have some confidence that the thing isn't just going to fall over once its been fully started.

I guess the nuisance here is that it is more difficult to run queries on the database before starting the reactor?

**Comment by NegativeMjark** [comment]
I think we already running queries on the database before the reactor has started to do the schema version checks before daemonising.

**Comment by dbkr** [comment]
Now modified to do checks before daemonising, with its own database conn.

**Comment by NegativeMjark** [comment]
LGTM

**Pull request merged**